File: platform/worker/src/generate.js
// Ad concept + persona generation — JS port of generate.py / personas.py.
import { llmComplete, fakeConcepts, fakePersonas, parseJsonArray } from "./providers.js";

export async function generateConcepts(cfg, brand, n = 6, opts = {}) {
  if (!cfg.__llm) return tag(fakeConcepts(n));
  const system =
    "You are a senior direct-response creative. Generate distinct, testable ad " +
    "concepts for paid social. Lead with the prospect/problem (never the brand), one " +
    "idea + one CTA each, specific over generic, match the awareness stage, stack real " +
    "persuasion. Vary the ANGLE (problem-led, founder POV, before/after, social proof, " +
    "mechanism, listicle). Return ONLY a JSON array; each object: " +
    '{"name","angle","awareness_stage","headline","primary_text","description"}.';
  const parts = [
    `BRAND: ${brand.name || ""}`, `URL: ${brand.url || ""}`,
    `WHAT IT IS: ${brand.description || ""}`, `TARGET CUSTOMER: ${brand.target_customer || ""}`,
  ];
  if (opts.references) parts.push(`\nCOMPETITOR / REFERENCE ADS (imitate the angles):\n${opts.references}`);
  if (opts.winners) parts.push(`\nOUR OWN PAST WINNERS (lean in):\n${opts.winners}`);
  if (opts.killList?.length) parts.push(`\nDO NOT REPEAT these failed angles: ${opts.killList.join("; ")}`);
  parts.push(`\nProduce exactly ${n} ad concepts as a JSON array.`);
  const concepts = parseJsonArray(await llmComplete(cfg, system, parts.join("\n"), 4000));
  return tag(concepts.slice(0, n));
}

function tag(concepts) {
  return concepts.map((c, i) => ({
    id: `c${i + 1}`, name: c.name || `Concept ${i + 1}`, angle: c.angle || "",
    awareness_stage: c.awareness_stage || "", headline: c.headline || "",
    primary_text: c.primary_text || "", description: c.description || "",
  }));
}

const STAGES = ["unaware", "problem-aware", "solution-aware", "product-aware", "most-aware"];

export async function generatePersonas(cfg, brand, n = 10) {
  if (!cfg.__llm) return fakePersonas(n);
  const system =
    "Generate a realistic, diverse panel of consumer personas for market research, " +
    `spread across the awareness stages (${STAGES.join(", ")}). Return ONLY a JSON array; ` +
    'each object: {"id","name","segment","awareness_stage","description"} where description ' +
    "is 1-2 sentences of demographics, motivation, and buying posture.";
  const user =
    `BRAND: ${brand.name || ""}\nWHAT IT IS: ${brand.description || ""}\n` +
    `TARGET CUSTOMER: ${brand.target_customer || ""}\n\nProduce exactly ${n} personas as a JSON array.`;
  const arr = parseJsonArray(await llmComplete(cfg, system, user, 4000));
  const out = (arr.length ? arr : [{}]).slice(0, n).map((p, i) => ({
    id: p.id || `p${i + 1}`, name: p.name || `Persona ${i + 1}`, segment: p.segment || "",
    awareness_stage: p.awareness_stage || STAGES[i % STAGES.length], description: p.description || "",
  }));
  return out;
}

File: platform/phase0/generate.py
"""Generate candidate ad concepts from a brand brief.

The engine could only *rank* ads it was given; this lets it *create* them. Output
is concept-level (headline + primary text + visual direction + angle) — exactly
what SSR scores. Rendering the visual to a finished image is a later slot
(the model router); the differentiator (SSR ranking) works at the concept level.
"""

from __future__ import annotations

import json
from typing import Dict, List, Optional

from providers import LLMProvider


def generate_concepts(
    llm: LLMProvider,
    brand: Dict,
    n: int = 6,
    references: Optional[str] = None,
    kill_list: Optional[List[str]] = None,
    winners: Optional[str] = None,
) -> List[Dict]:
    """Produce ``n`` distinct ad concepts grounded in the brand + references."""
    system = (
        "You are a senior direct-response creative. Generate distinct, testable ad "
        "concepts for paid social. Apply the canon: lead with the prospect/problem "
        "(never the brand), one idea + one CTA per ad, specific over generic, match "
        "the awareness stage, stack real persuasion. Vary the ANGLE across concepts "
        "(problem-led, founder POV, before/after, social proof, mechanism, "
        "listicle). Return ONLY a JSON array; each object: "
        '{"name","angle","awareness_stage","headline","primary_text","description"}. '
        '"description" is the visual direction for the image.'
    )
    parts = [
        f"BRAND: {brand.get('name','')}",
        f"URL: {brand.get('url','')}",
        f"WHAT IT IS: {brand.get('description','')}",
        f"TARGET CUSTOMER: {brand.get('target_customer','')}",
    ]
    if references:
        parts.append(f"\nCOMPETITOR / REFERENCE ADS THAT WORK (imitate the angles):\n{references}")
    if winners:
        parts.append(f"\nOUR OWN PAST WINNERS (lean into these):\n{winners}")
    if kill_list:
        parts.append("\nDO NOT REPEAT these angles/hooks (they failed): " + "; ".join(kill_list))
    parts.append(f"\nProduce exactly {n} ad concepts as a JSON array.")
    raw = llm.complete(system, "\n".join(parts), max_tokens=4000)

    concepts = _parse_json_array(raw)
    out: List[Dict] = []
    for i, c in enumerate(concepts[:n]):
        out.append(
            {
                "id": f"c{i+1}",
                "name": c.get("name") or f"Concept {i+1}",
                "angle": c.get("angle", ""),
                "awareness_stage": c.get("awareness_stage", ""),
                "headline": c.get("headline", ""),
                "primary_text": c.get("primary_text", ""),
                "description": c.get("description", ""),
            }
        )
    return out


def _parse_json_array(raw: str) -> List[Dict]:
    raw = raw.strip()
    start, end = raw.find("["), raw.rfind("]")
    if start == -1 or end == -1:
        return []
    try:
        data = json.loads(raw[start : end + 1])
        return data if isinstance(data, list) else []
    except json.JSONDecodeError:
        return []
